src/CRISP.BackendChallenge/Controllers/EmployeeController.cs

```diff
@@ -7,33 +7,266 @@
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(":: Error occurred while searching employees: {EMessage}", e.Message);
+            return CreateInternalServerError();
+        }
+        
+    }
+    
+    
+    [HttpPut]
+    [Route("update/{id:int}")]
+    public IActionResult UpdatedById(int id,[FromBody] EmployeeRequest employeeRequest)
+    {
+        _logger.LogDebug(":: Performing {MethodName} with ID {Id}", nameof(UpdatedById), id);
+        
+        try
+        {
+           
+            var existingEmployee = _employeeRepository.GetById(id);
+            
+            existingEmployee.Name = employeeRequest.Name ?? existingEmployee.Name;
+            if (employeeRequest.Department != null)
             {
-                Id = x.Id,
-                Name = x.Name,
-                // TODO: Include the login date information...
-                LoginDates = null
-            });
-        return Ok(result);
-    }
-    // TODO: Implement Search By Id
-    // TODO: Implement Delete by Id
-    // TODO: Implement Update by Id
+                existingEmployee.Department = (Department)employeeRequest.Department;
+            }
+            existingEmployee.StartDate = employeeRequest.StartDate;
+            existingEmployee.EndDate = employeeRequest.EndDate;
+            
+            _employeeRepository.Update(existingEmployee);
+            _employeeRepository.Save();
+            
+            if (employeeRequest.LoginDates != null)
+            {
+                foreach (var login in employeeRequest.LoginDates.Select(l => new Login()
+                         {
+                             EmployeeId = existingEmployee.Id,
+                             LoginDate = l
+                         }))
+                {
+                    _loginRepository.Add(login);
+                }
+                _loginRepository.Save();
+            }
+
+            return Ok(ToEmployeeResponse(existingEmployee));
+        }
+        catch (ArgumentNullException e)
+        {
+            _logger.LogDebug(":: Employee Not Found: {Id}, {EMessage}", id, e.Message);
+            return CreateNotFound(id);
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(":: Error occurred while updating employee with ID {Id}: {EMessage}", id, e.Message);
+            return CreateInternalServerError();
+        }
+    }
+    
+
+    [HttpDelete]
+    [Route("delete/{id:int}")]
+    public IActionResult DeleteById(int id)
+    {
+        _logger.LogDebug(":: Performing {MethodName} with ID {Id}", nameof(DeleteById), id);
+        try
+        {
+            var employee = _employeeRepository.GetById(id);
+            
+            
+            var loginDates = _loginRepository.Query().Where(l => l.EmployeeId == id).ToList();
+            foreach (Login login in loginDates)
+            {
+                _loginRepository.Delete(login);
+                _loginRepository.Save();
+            }
+
```

Comment:
SQL server provides this functionality out of the box with cascade delete mode - I would leverage that instead.

Outside of that - you should pretty much NEVER call `.Save()` or `.SaveAsync()` more than once within the scope of a single repository method (which is what this should be in), barring very specific circumstances.

In this instance, you're deleting each single login with a separate request to server, rather than all at once.

Reply:
Updated the delete method to cascade delete instead along with cleaning up the `.Save()` around in the controller